### PR TITLE
Create admin API to list cluster members

### PR DESCRIPTION
**Pull request by feedmeapples**
**What changed?**

Added an admin API to get cluster members

**Why?**

exposes cluster members as admin API. Removes a need for direct persistence connection in tctl

**How did you test it?**

- tested with `tctl admin membership list_db`

**Potential risks**

afaik no added risks, only possible errors with the new API itself 

**Is hotfix candidate?**

no

### CODE REVIEW

File: proto/internal/temporal/server/api/adminservice/v1/request_response.proto [hunk 1]
```diff
@@ -201,6 +203,25 @@ message DescribeClusterResponse {
     temporal.api.version.v1.VersionInfo version_info = 9;
 }
 
+message GetClusterMembersRequest {
+    // (-- api-linter: core::0140::prepositions=disabled
+    //     aip.dev/not-precedent: "within" is used to indicate a time range. --)
+    google.protobuf.Duration last_heartbeat_within = 1 [(gogoproto.stdduration) = true];
```
Comment (feedmeapples) [inline proto/internal/temporal/server/api/adminservice/v1/request_response.proto:209]: any ideas how to replace the "within" part that is getting linted?

File: service/frontend/adminHandler.go [hunk 2]
```diff
@@ -672,6 +672,42 @@ func (adh *AdminHandler) DescribeCluster(_ context.Context, _ *adminservice.Desc
...
+	scope, sw := adh.startRequestProfile(metrics.AdminGetClusterMembersScope)
+	defer sw.Stop()
+
+	if request == nil {
+		return nil, adh.error(errRequestNotSet, scope)
+	}
+
+	metadataMgr := adh.GetClusterMetadataManager()
+
+	resp, err := metadataMgr.GetClusterMembers(&persistence.GetClusterMembersRequest{NextPageToken: nil})
```
Comment (yux0) [inline service/frontend/adminHandler.go:690]: Do we plan to wire the input request and the persistence request?

Comment (feedmeapples) [reply service/frontend/adminHandler.go:690]: good catch!

Comment (yiminc) [reply service/frontend/adminHandler.go:690]: what is the default page size? do we need to be worried about more clusters than page size?

File: service/frontend/adminHandler.go [hunk 3]
```diff
@@ -672,6 +672,42 @@ func (adh *AdminHandler) DescribeCluster(_ context.Context, _ *adminservice.Desc
...
+			RpcPort:          int32(member.RPCPort),
+			SessionStartTime: &member.SessionStart,
+			LastHeartbitTime: &member.LastHeartbeat,
+			RecordExpiryTime: &member.RecordExpiry,
+		})
+	}
+
+	return &adminservice.GetClusterMembersResponse{
+		ActiveMembers: activeMembers,
+		NextPageToken: resp.NextPageToken,
```
Comment (yiminc) [inline service/frontend/adminHandler.go:707]: do you expect caller to page through? if so, i don't see you set the page token from input request

File: proto/internal/temporal/server/api/cluster/v1/message.proto [hunk 4]
```diff
@@ -39,3 +44,13 @@ message MembershipInfo {
...
 }
+
+message ClusterMember {
+    temporal.server.api.enums.v1.ClusterMemberRole role = 1;
+    string host_id = 2;
+    string rpc_address = 3;
+    int32 rpc_port = 4;
+    google.protobuf.Timestamp session_start_time = 5 [(gogoproto.stdtime) = true];
+    google.protobuf.Timestamp last_heartbit_time = 6 [(gogoproto.stdtime) = true];
+    google.protobuf.Timestamp record_expiry_time = 7 [(gogoproto.stdtime) = true];
```
Comment (yiminc) [inline proto/internal/temporal/server/api/cluster/v1/message.proto:55]: are these coming from ringpop?

File: service/frontend/adminHandler.go (hunk 2 above)
Comment (feedmeapples) [reply service/frontend/adminHandler.go:690]: @yiminc yes we need to take this into account, luckily there is already pagination supported. Just need to wire the nextPageToken to the persistence request

File: service/frontend/adminHandler.go (hunk 3 above)
Comment (feedmeapples) [reply service/frontend/adminHandler.go:707]: afaict this is the same issue as mentioned here? https://github.com/temporalio/temporal/pull/2159/files/ba343e7acd0edf9e6914b272f481c18deed7d4f4#r747687170

addressing it

File: proto/internal/temporal/server/api/cluster/v1/message.proto (hunk 4 above)
Comment (feedmeapples) [reply proto/internal/temporal/server/api/cluster/v1/message.proto:55]: this comes from `cluster_membership` table, not ringpop (as i understand this) https://github.com/temporalio/temporal/blob/590b6eaeb0f183216290ac24249abee04a9716d3/common/persistence/cassandra/cluster_metadata_store.go#L136

Comment (feedmeapples) [reply proto/internal/temporal/server/api/cluster/v1/message.proto:55]: sample output

```
[
  {
    "role": 2,
    "host_id": "0235beb1-4265-11ec-88f8-50e085dc7c06",
    "rpc_address": "127.0.0.1",
    "rpc_port": 6934,
    "session_start_time": "2021-11-10T20:30:13.829851Z",
    "last_heartbit_time": "2021-11-10T22:49:11.729838Z",
    "record_expiry_time": "2021-11-12T22:49:11.729838Z"
  },
  {
    "role": 3,
    "host_id": "02374d05-4265-11ec-88f8-50e085dc7c06",
    "rpc_address": "127.0.0.1",
    "rpc_port": 6935,
    "session_start_time": "2021-11-10T20:30:13.829859Z",
    "last_heartbit_time": "2021-11-10T22:49:11.729886Z",
    "record_expiry_time": "2021-11-12T22:49:11.729886Z"
  },
  ```

File: common/metrics/defs.go [hunk 5]
```diff
@@ -1393,6 +1397,7 @@ var ScopeDefs = map[ServiceIdx]map[int]scopeDefinition{
 		AdminGetWorkflowExecutionRawHistoryScope:   {operation: "GetWorkflowExecutionRawHistory"},
 		AdminGetWorkflowExecutionRawHistoryV2Scope: {operation: "GetWorkflowExecutionRawHistoryV2"},
 		AdminGetReplicationMessagesScope:           {operation: "GetReplicationMessages"},
+		AdminGetClusterMembersScope:                {operation: "AdminGetClusterMembersScope"},
```
Comment (feedmeapples) [inline common/metrics/defs.go:1400]:
```suggestion
		AdminGetClusterMembersScope:                {operation: "GetClusterMembers"},
```

File: common/metrics/defs.go [hunk 6]
```diff
@@ -1393,6 +1397,7 @@ var ScopeDefs = map[ServiceIdx]map[int]scopeDefinition{
 		AdminGetWorkflowExecutionRawHistoryScope:   {operation: "GetWorkflowExecutionRawHistory"},
 		AdminGetWorkflowExecutionRawHistoryV2Scope: {operation: "GetWorkflowExecutionRawHistoryV2"},
 		AdminGetReplicationMessagesScope:           {operation: "GetReplicationMessages"},
+		AdminGetClusterMembersScope:                {operation: "GetClusterMembers"},
```
Comment (feedmeapples) [inline common/metrics/defs.go:1400]:
```suggestion
		AdminGetClusterMembersScope:                {operation: "AdminGetClusterMembers"},
```

File: service/frontend/adminHandler.go (hunk 2 above)
Comment (feedmeapples) [reply service/frontend/adminHandler.go]: addressed

File: service/frontend/adminHandler.go (hunk 3 above)
Comment (feedmeapples) [reply service/frontend/adminHandler.go:728]: addressed

**Review by yiminc — APPROVED** [review]

**Pull request merged**